WebAPI/Controllers/ExchangeController.cs

```diff
@@ -7,49 +7,86 @@
+        private readonly IUserService _userService;
+        private readonly IStudentService _studentService;
+        private readonly Uri _baseUrl;
 
-        public ExchangeController(IBlockChangesService blockChangeService, INotificationService notificationService) : base()
+        public ExchangeController(ILogger<UserController> logger, IBlockChangesService blockChangeService, IEmailService emailService,
+            IUserService userService, IStudentService studentService, IOptions<EnvironmentSettings> environmentSettings) : base()
         {
+            _logger = logger;
             _blockChangesService = blockChangeService;
-            _notificationService = notificationService;
+            _emailService = emailService;
+            _userService = userService;
+            _studentService = studentService;
+            _baseUrl = new Uri(environmentSettings.Value.BaseUrl);
         }
 
         [HttpPost]
         [Route("exchangeConfirm")]
+        [AllowAnonymous]
         public async Task<IActionResult> ExchangeConfirm([FromBody]ExchangeRequestModel request)
         {
-            var currentUserBlockChangeRequest = new BlockChangeRequest();
-            currentUserBlockChangeRequest.BlockFrom = BlockForExchangeModel.ConvertToBlock(request.BlockFrom);
-            currentUserBlockChangeRequest.BlockTo = BlockForExchangeModel.ConvertToBlock(request.BlockTo);
-            currentUserBlockChangeRequest.Status = ExchangeStatus.WaitingForExchange;
-            currentUserBlockChangeRequest.DateOfCreation = DateTime.Now;
-            currentUserBlockChangeRequest.StudentId = Guid.Parse(request.StudentId);
+            var blockChangeRequest = new BlockChangeRequest();
+            blockChangeRequest.BlockFrom = BlockForExchangeModel.ConvertToBlock(request.BlockFrom);
+            blockChangeRequest.BlockTo = BlockForExchangeModel.ConvertToBlock(request.BlockTo);
+            blockChangeRequest.Status = ExchangeStatus.WaitingForExchange;
+            blockChangeRequest.DateOfCreation = DateTime.Now;
+            blockChangeRequest.StudentId = Guid.Parse(request.StudentId);
 
-            var (exchangeWasMade, foundMatch) = await _blockChangesService.AddAndFindMatch(currentUserBlockChangeRequest);
-
-            if (exchangeWasMade)
+            var res = await _blockChangesService.AddAndFindMatch(blockChangeRequest);
+            if (res != null)
             {
-                await _notificationService.AddNotification(
-                    SuccessfulExchangeNotification.Create(currentUserBlockChangeRequest, foundMatch));
+                var student1 = await _studentService.FindByIdAsync(Guid.Parse(res.FirstID));
+                var user1 = await _userService.GetUserByIdAsync(student1.UserId.ToString());
+                if (user1 == null)
+                {
+                    _logger.LogError($"Cannot find user with ID {res.FirstID}.");
+                    return BadRequest();
```

Comment:
BadRequest - but exchange was realized... Is it possible to solve it somehow else... or this check has no sence..
May be the notification should be part of BlockChangeService